<a href="https://colab.research.google.com/github/ShivangiLab/Predictive-Analysis--Linear-Regression-on-California-Housing-data./blob/main/Shivangi_Jaidka_Lab01_BAN210.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **Question 1: Import Libraries and Load Dataset [2 Marks]**

In [ ]:
# Importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.datasets import fetch_california_housing
# Import necessary libraries you need for rest of lab below

## TODO

# Splits data into train/test
from sklearn.model_selection import train_test_split

 # The Linear Regression model
from sklearn.linear_model import LinearRegression

# Evaluation metrics
from sklearn.metrics import mean_squared_error, r2_score

# Load the dataset
data = fetch_california_housing()
X = pd.DataFrame(data.data, columns=data.feature_names)
y = pd.Series(data.target, name="Target")

# Display first few rows of the dataset
## TODO
# this will give the first five rows starts from 0-4 of the features

(X.head())

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25


The California Housing dataset contains information about housing districts in California. The dataset includes 8 predictor variables such as median income, house age, average number of rooms, population, occupancy, latitude, and longitude.

The target variable represents the median house value for each district. The objective of this analysis is to build a Linear Regression model that can predict housing prices based on these features


### **Question 2: Train-Test Split [1 Marks]**

In [ ]:
X = pd.DataFrame(data.data, columns=data.feature_names)
y = pd.Series(data.target, name="Target")

# TODO: Split the data into training and test sets (80% train, 20% test)

#80% of the data will be used to train and 20% of the data will be used to test
#so that it gives us a good balance of the evaluation.

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# DO NOT CHANGE BELOW CODE
print(f"Training data size: {X_train.shape}")
print(f"Test data size: {X_test.shape}")


Training data size: (16512, 8)
Test data size: (4128, 8)


The dataset was divided into training and testing sets using an 80-20 split.

- Training Data: 16,512 observations
- Testing Data: 4,128 observations

The training set is used to teach the model the relationship between the features and housing prices, while the testing set is used to evaluate the model's performance on unseen data. An 80-20 split provides a good balance between model training and evaluation.

### **Question 3: Implement Linear Regression [2 Marks]**

In [ ]:
### START CODE HERE ### (≈ 3 lines of code)

# This cell focuses on creating the Linear Regression model and training
#it with our training data.

# TODO: Initialize the LinearRegression model

# Initialize a Linear Regression model to predict housing prices

model = LinearRegression()

# TODO: Train the model using X_train and y_train

# Fit the model to the training data so it can learn relationships
# between the housing features and the target variable

model.fit(X_train, y_train)

### END CODE HERE ###

# DO NOT CHANGE BELOW CODE
print("Linear Regression model trained.")


Linear Regression model trained.


## Linear Regression Model

A Linear Regression model was selected because it is one of the most commonly used supervised learning algorithms for predicting continuous numerical values.

The model learns the relationship between the housing features and the target variable by fitting a linear equation to the training data.

### **Question 4: Evaluation - Mean Squared Error (MSE) [3 Marks]**

In [ ]:
### START CODE HERE ### (≈ 4 lines of code)

# TODO: Predict on the training set
y_pred_train = model.predict(X_train)

# TODO: Predict on the test set
y_pred_test = model.predict(X_test)

# TODO: Calculate and print the MSE for training and test sets

#Measures the average of the squares of the errors.
# Evaluate model accuracy using Mean Squared Error (MSE).
#MSE= Measures the average of the squares of the errors,
#i.e., the average squared difference between the estimated
#values and actual value.

mse_train = mean_squared_error(y_train, y_pred_train)
mse_test = mean_squared_error(y_test, y_pred_test)

### END CODE HERE ###

# DO NOT CHANGE BELOW CODE
print(f"MSE on training set: {mse_train}")
print(f"MSE on test set: {mse_test}")


MSE on training set: 0.5179331255246699
MSE on test set: 0.5558915986952444


## Mean Squared Error (MSE)

The Mean Squared Error measures the average squared difference between the actual and predicted housing prices.

Results:
- Training MSE = 0.518
- Testing MSE = 0.556

The testing MSE is only slightly higher than the training MSE, indicating that the model performs consistently on both datasets. This suggests that the model is not significantly overfitting and is able to generalize reasonably well to unseen data.


### **Question 5: Evaluation - R-squared (R²) [2 Marks]**

In [ ]:
### START CODE HERE ### (≈ 2 lines of code)

# TODO: Calculate and print the R² score for training set

# R² measures how much of the variance in
#house prices is explained by our model's features

r2_train = r2 = r2_score(y_train, y_pred_train)

# TODO: Calculate and print the R² score for test set

# Comparing training and testing R² scores helps determine
# how well the model generalizes to unseen data

r2_test = r2 = r2_score(y_test, y_pred_test)

### END CODE HERE ###

# DO NOT CHANGE BELOW CODE
print(f"R² on training set: {r2_train}")
print(f"R² on test set: {r2_test}")


R² on training set: 0.6125511913966952
R² on test set: 0.5757877060324508


## R² Score Analysis

The R² score measures the proportion of variation in housing prices that can be explained by the predictor variables.

Results:
- Training R² = 0.613
- Testing R² = 0.576

The model explains approximately 61.3% of the variation in housing prices on the training data and 57.6% on the testing data. The relatively small difference between these values suggests that the model generalizes well and does not suffer from severe overfitting.

## Model Evaluation

The Linear Regression model achieved moderate predictive performance on the California Housing dataset.

Key observations:
- Training and testing MSE values are relatively close.
- Training and testing R² values are also similar.
- The model is able to explain more than half of the variation in housing prices.

These results indicate that the model has learned meaningful relationships from the data and performs reasonably well when making predictions on unseen observations.

## Conclusion

A Linear Regression model was successfully developed and evaluated using the California Housing dataset.

The model achieved:
- Training MSE: 0.518
- Testing MSE: 0.556
- Training R²: 0.613
- Testing R²: 0.576

Based on these results, the model demonstrates reasonable predictive capability and generalizes well to new data. While it does not explain all variation in housing prices, it provides a solid baseline model for housing price prediction and highlights the usefulness of Linear Regression for predictive analytics tasks.